src/main/java/software/amazon/ionpathextraction/PathExtractorImpl.java

```diff
@@ -145,24 +149,47 @@ private int invokeCallback(final IonReader reader,
         return stepOutTimes;
     }
 
-    private boolean pathComponentMatches(final SearchPath<T> searchPath,
-                                         final IonReader reader,
-                                         final int currentDepth,
-                                         final int currentPosition) {
-        List<PathComponent> pathComponents = searchPath.getPathComponents();
-
-        // currentDepth 0 can only match the empty search path: ()
-        if (currentDepth == 0) {
-            return pathComponents.isEmpty();
-        } else if (currentDepth <= pathComponents.size()) {
-            return pathComponents.get(currentDepth - 1).matches(reader, currentPosition, config);
+    private boolean searchPathPartialMatchAt(final SearchPath<T> searchPath,
+                                             final IonReader reader,
+                                             final int currentDepth,
+                                             final int currentPosition) {
+        switch (searchPath.getType()) {
+            case TOP_LEVEL:
+                return currentDepth == 0;
+
+            case ANNOTATED_TOP_LEVEL:
+                AnnotatedTopLevelSearchPath<T> annotatedSearchPath = (AnnotatedTopLevelSearchPath<T>) searchPath;
+                return currentDepth == 0
+                    && arrayEquals(
+                        reader.getTypeAnnotations(),
```

Comment:
Note: this comment is about potential performance implications that don't necessarily need to be solved in this PR. Creating an issue to investigate further would be sufficient for now.

`searchPathPartialMatchAt` is called for every active path at a given reader location. This location will have the same annotations on each call, but `reader.getTypeAnnotations()` is called every time, which allocates a brand new array unless the location has zero annotations. We could consider storing the annotation array once and passing it in. The same can be said about `AnnotatedPathComponent.matches`.